Zend/zend_compile.c

```diff
@@ -1263,7 +1263,8 @@ static void zend_mark_function_as_generator() /* {{{ */
 
 	if (CG(active_op_array)->fn_flags & ZEND_ACC_HAS_RETURN_TYPE) {
 		zend_type return_type = CG(active_op_array)->arg_info[-1].type;
-		zend_bool valid_type = (ZEND_TYPE_FULL_MASK(return_type) & MAY_BE_ITERABLE) != 0;
+		zend_bool valid_type = (ZEND_TYPE_FULL_MASK(return_type) & MAY_BE_ITERABLE) != 0 ||
+			ZEND_TYPE_FULL_MASK(return_type) == MAY_BE_ANY;
```

Comment:
This should actually check for `& (MAY_BE_ITERATABLE|MAY_BE_OBJECT) != 0`. `: object` is also a fine type.

Reply:
At which point we should really adjust the error message to be generic and not list every single possible type. Something along the lines of...

> Cannot use %s as generator return type, because it is not a supertype of Generator

> Generator return type must be a supertype of Generator, %s given

Reply:
Done